### PR TITLE
Cleaning stale validation metrics collected during sanity check from rich progress bar

### PR DESCRIPTION
**Pull request by tesfaldet**
## What does this PR do?

Fixing issue where training starts with stale validation metrics collected during sanity check shown in the rich progress bar. Added additional resetting of validation metrics.

## Before submitting

- [x] Did you make sure **title is self-explanatory** and **the description concisely explains the PR**?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **test your PR locally** with `pytest` command?
- [x] Did you **run pre-commit hooks** with `pre-commit run -a` command?

## Did you have fun?

Mhm :)

### CODE REVIEW

File: src/models/mnist_module.py [hunk 1]
```diff
@@ -59,6 +60,16 @@ def on_train_start(self):
         # so we need to make sure val_acc_best doesn't store accuracy from these checks
         self.val_acc_best.reset()
 
+        # although val_loss and val_acc are automatically reset at end of sanity check (since they're fed as
+        # torchmetrics.Metric objects to self.log in validation_step with on_epoch=True), reset them anyways
+        self.val_loss.reset()
+        self.val_acc.reset()
```
Comment (ashleve) [inline src/models/mnist_module.py:66]: I'm confused by the reasoning here. Is resetting all validation metrics necessary before the progress bar fix later on?

**Comment by ashleve** [comment]
> training starts with stale validation metrics collected during sanity check shown in the rich progress bar

 I am having some difficulty understanding the issue. Is this effect visible on the mnist example? When I launch the training, the sanity checks go so fast that I don't notice them in the progress bar. Maybe this is something you experience with slower models?

File: src/models/mnist_module.py (hunk 1 above)
Comment (ashleve) [reply src/models/mnist_module.py:66]: Athough we could encourage resetting all validiation metrics at the start of the training like this:
```python
# by default lightning executes validation step sanity checks before training starts,
# so we need to make sure validation metrics don't store results from these checks
self.val_acc_best.reset()
self.val_loss.reset()
self.val_acc.reset()
```

(just in case it will save someone trouble when doing something unconventional with validation metrics)

**Comment by tesfaldet** [comment]
I realized I should have provided screenshots. I'm on my way to work but I'll share screenshots later today.

File: src/models/mnist_module.py (hunk 1 above)
Comment (tesfaldet) [reply src/models/mnist_module.py:66]: Yes, the suggestion above makes more sense. It's what I originally would have put so as to encourage resetting of all validation metrics, even though val_loss and val_acc are in fact resetted.

To answer your Q: no, it's not necessary to reset all validation metrics for the progress bar fix. However, instead of explaining why that's the case to the user—which is long-winded and nuanced (as you saw in my code comment above)—it's probably better to just encourage explicitly resetting any validation metrics that were logged in either `validation_step()` or `validation_epoch_end()` callbacks.

**Comment by tesfaldet** [comment]
I've got some screenshots to better explain what's going on:

## Without fix

Sanity check has ended, `validation_epoch_end()` is called.

Screenshot 2023-02-24 at 09 44 33

Training has started. The sanity check progress bar has been removed by Lightning. More accurately, it has been set to be [invisible](https://github.com/Lightning-AI/lightning/blob/6b9ddf00dd0a6c41276491641f927147bfedda58/src/lightning/pytorch/callbacks/progress/rich_progress.py#L359). It still exists, and thus the `MetricsTextColumn` [object](https://github.com/Lightning-AI/lightning/blob/6b9ddf00dd0a6c41276491641f927147bfedda58/src/lightning/pytorch/callbacks/progress/rich_progress.py#L137-L174) that's used to display the user-tracked metrics (those fed into `self.log()`) inside the `RichProgressBar` object still contains values inside its `_metrics` property (which is a dictionary containing metric names and values to be displayed). This is still the case despite resetting the validation `Metric` objects, as those are not what's stored inside `MetricsTextColumn._metrics`. Rather, it's their most previous *computed* values that's stored in that property.

Screenshot 2023-02-24 at 09 44 51

The first training step has begun. The progress bar is displayed. `loss` is showing `nan` as expected but you can see that the most recently computed user-tracked validation metrics, `val/acc` and `val/acc_best`, are shown (`0.133`).

Screenshot 2023-02-24 at 09 45 05

## With fix

By explicitly calling `self.trainer.progress_bar_callback._update_metrics(self.trainer, self)` we are forcing a [refresh](https://github.com/Lightning-AI/lightning/blob/6b9ddf00dd0a6c41276491641f927147bfedda58/src/lightning/pytorch/callbacks/progress/rich_progress.py#L565-L568) of the `RichProgressBar` object's `_metric_component` [property](https://github.com/Lightning-AI/lightning/blob/6b9ddf00dd0a6c41276491641f927147bfedda58/src/lightning/pytorch/callbacks/progress/rich_progress.py#L265) which contains the user-tracked computed validation metrics from the most recent validation step. This will retrieve the currently tracked metrics (which have all been resetted at this point) and update the progress bar's stored metrics.

Screenshot 2023-02-24 at 09 48 23

So yeah, basically this is an issue on Lightning's end and the fix is a way to deal with it. Ideally, Lightning would do the resetting properly. Fortunately, this has no negative consequence that I can think of, but I feel that it's an unintended side-effect that could lead to downstream issues with the progress bar, hence the fix over here.

**Comment by ashleve** [comment]
Thank you for the visual explanation! I understand the issue but it's interesting because I don't experience the metrics from sanity checks being shown. Here's my terminal in vscode on macos zsh:

image

As you can there's only loss being shown - the rest of the metrics appear only after first epoch is finished.

What system and setup are you using?

If this is issue with Lightning then I would prefer to send them a PR instead of introducing fixes to the template. I would be glad to take care of this if you don't have the bandwidth but I haven't been able to reproduce the issue so far.

**Comment by tesfaldet** [comment]
The issue only shows itself during the very first training step. Can you put a breakpoint on the first line of `training_step()` and show me the state of the progress bar when the breakpoint is hit for the first time? What I'm seeing in your screenshot is at step 318, which at that point you won't see those stale metrics. What I'm trying to say is that the stale validation metrics exist for the moment between sanity checking has ended and when training has just begun. Specifically, right when the training progress bar is first displayed but the first train step hasn't finished.

**Comment by tesfaldet** [comment]
> If this is issue with Lightning then I would prefer to send them a PR instead of introducing fixes to the template. I would be glad to take care of this if you don't have the bandwidth but I haven't been able to reproduce the issue so far.

Ah no worries, I was already planning to send them a PR. I think it might make more sense than to introduce a fix here since it has nothing to do with the template, as far as I can think of. So perhaps we can close this, I can try to reproduce this issue in other Lightning-based project setups, then I'll introduce the bug to the Lightning folks.

**Comment by ashleve** [comment]
I see. It does indeed appear at first step. I've never considered this an issue since this is visible only for a short moment, but you're right it might be worth to reset the progress bar after validation sanity checks just in case there might be some unintended consequences. 

Personally I don't feel like it's worth expanding the LightningModule code just for the small visual fix.

If you're planning on sending PR to lightning library then I'm closing this for now. I could also go for the PR myself so let me know if you change your mind.

**Comment by tesfaldet** [comment]
>Personally I don't feel like it's worth expanding the LightningModule code just for the small visual fix.

Haha yeah I agree. You make a good point. I'll try to go for the PR sometime this week and will let you know if I don't have enough bandwidth for it. Thanks!

File: src/models/mnist_module.py (hunk 1 above)
Comment (ashleve) [reply src/models/mnist_module.py:66]: Regarding the reset of all validation metrics when training starts - thank you for this suggestion, I've made appriopriate PR here: https://github.com/ashleve/lightning-hydra-template/pull/540
